### PR TITLE
IIIF Viewer Refactor Tests phase B

### PR DESCRIPTION
**Pull request by davidpmccormick**
- For #12984

## What does this change?

Adds characterisation tests locking in the current behaviour of the `utils/iiif/v3` functions that feed the viewer's derived state, so the upcoming context refactor can be verified green-to-green. Extends two existing suites (`index.test.ts`, `canvas.test.ts`) — no source/behaviour changes.

Coverage added:
- **Restriction:** `isItemRestricted`, `hasRestrictedItem`, `getProbeServiceId`, `checkModalRequired`
- **Downloads:** `getDownloadOptionsFromManifestRendering`, `getDownloadOptionsFromCanvasRenderingAndSupplementing`, `getVideoAudioDownloadOptions`
- **Media-type / layout selection:** `hasItemType`, `hasOriginalPdf`, `isPDFCanvas`, `isAudioCanvas`, `getItemsStatus`, `getFileTypeLabel`
- **Helpers:** `isChoiceBody`, `getImageServiceFromItem`, `getFileSize`, `getOriginalFiles`, `isCollection`
- **Integration:** `transformCanvas`; plus `getOriginal` and `getDisplayItems` in `canvas.test.ts

Notes for reviewers:
- These are **characterisation** (not specification) tests: they assert what the code *currently* does, quirks included. One quirk is pinned with a comment rather than "fixed" — `getFileTypeLabel(0, 0, false, [])` returns `"0 PDF files"` because an empty array makes `every(isPDFCanvas)` vacuously true (`undefined` canvases give `"0 images"`).
- **Not exhaustive by design** — scope is the derived-state surface the refactor touches. Deliberately out of scope for now: the auth-transformer layer, `groupRanges`, metadata extractors, and label helpers (some exercised indirectly via `transformLabel`/`transformCanvas`). The remaining RFC derived-value/boolean checklist (navigation state, `mainImageService` fallback, image-service flags) is inline in components and will be added in a subsequent PR.

## How to test

1. `test:content` — full content suite passes (95 tests across `utils/iiif/v3` + harness)
2. `yarn workspace @weco/content test:watch:fast iiif` — confirms the new suites run under the fast watch loop added in #13245
3. `yarn tsc` and `yarn lint`

## Have we considered potential risks?

There should be no behaviour changes, just adding tests, so I think we're ok.

### CODE REVIEW

File: content/webapp/package.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "dev": "NODE_EXTRA_CA_CERTS=\"$(mkcert -CAROOT)/rootCA.pem\" NODE_ENV=development tsx watch --tsconfig tsconfig.server.json server.ts",
     "start": "NODE_ENV=production tsx --tsconfig tsconfig.server.json server.ts",
     "test": "NODE_ENV=test jest --no-cache",
-    "test:watch": "yarn test -- --watchAll"
+    "test:watch": "yarn test -- --watchAll",
+    "test:watch:fast": "NODE_ENV=test jest --watch"
   },
```
[the remark this answers is not in the document]
Comment (davidpmccormick) [reply content/webapp/package.json:12]: 🤓

[the remark this answers is not in the document]
Comment (davidpmccormick) [reply content/webapp/package.json:12]: changed base branch

File: content/webapp/test/fixtures/iiif/render.test.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+const Probe = () => {
+  const { transformedManifest } = useItemViewerContext();
+  const { isFullSupportBrowser } = useAppContext();
+  const { userIsStaffWithRestricted } = useUserContext();
+  return (
+    <div>
+      <span data-testid="canvas-count">
+        {transformedManifest?.canvases.length}
+      </span>
```
[the remark this answers is not in the document]
Comment (davidpmccormick) [reply content/webapp/test/fixtures/iiif/render.test.tsx:21]: > One nuance worth stating precisely: in the **current** multi-line formatting, the JSX transform actually strips whitespace-only text that sits adjacent to a tag and contains a newline, so `<span>\n {expr}\n</span>` compiles to `createElement('span', …, expr)` with no whitespace text node — meaning `toBeEmptyDOMElement()` passes reliably *today*. So it's not flaky as written.
>
> But the reviewer's underlying point stands: the assertion's correctness hinges on that subtle JSX rule, so it's fragile to any reformatting and vague about intent. Better to not depend on it at all — render an explicit sentinel and assert on it. That also matches the RFC's own convention (`toHaveTextContent('none')` for an empty value)

File: content/webapp/test/fixtures/iiif/transformed-manifest.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+export function createMockManifest(
+  overrides: Partial<TransformedManifest> = {}
+): TransformedManifest {
+  const { canvases: overrideCanvases, auth: overrideAuth, ...restOverrides } =
+    overrides;
+
+  const canvases = overrideCanvases ?? [createMockCanvas()];
+  const auth = overrideAuth ?? { ...defaultMockAuth };
+
```
[the remark this answers is not in the document]
Comment (davidpmccormick) [reply content/webapp/test/fixtures/iiif/transformed-manifest.ts:115]: Replaced the shared `defaultMockAuth` const with a `createMockAuth(overrides?)` factory, so every manifest gets its own auth objectand its own `accessRequirements` array. No shared mutable state between factory outputs

File: content/webapp/test/fixtures/iiif/render.test.tsx [hunk 4]
```diff
@@ -40,7 +40,7 @@ describe('renderWithContext harness', () => {
       appContext: { isFullSupportBrowser: true },
       userContext: { userIsStaffWithRestricted: true },
     });
-    expect(screen.getByTestId('canvas-count')).toBeEmptyDOMElement();
+    expect(screen.getByTestId('canvas-count')).toHaveTextContent('none');
```
Comment (rcantin-w) [inline content/webapp/test/fixtures/iiif/render.test.tsx:47]: Why this change?

File: content/webapp/utils/iiif/v3/canvas.test.ts [hunk 5]
```diff
@@ -44,3 +66,57 @@ describe('getThumbnailImage', () => {
...
+      { id: 'other', behavior: 'something-else', type: 'Image' },
+      { id: 'no-behavior', type: 'Image' },
+    ];
+    expect(getOriginal(rendering as never).map(o => o.id)).toEqual([
+      'original',
+    ]);
+  });
+
+  it('returns an empty array when there are no originals', () => {
+    expect(getOriginal([])).toEqual([]);
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/canvas.test.ts:89]: This is fine, but wonder if it'd be nicer to test it with a non empty object

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 6]
```diff
@@ -1,12 +1,39 @@
...
+  hasOriginalPdf,
+  hasRestrictedItem,
+  isAudioCanvas,
+  isChoiceBody,
+  isCollection,
+  isItemRestricted,
+  isPDFCanvas,
+  transformCanvas,
   transformLabel,
 } from '.';
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:38]: dang 😄

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 7]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+      ],
+    });
+    expect(isItemRestricted(painting)).toBe(false);
+  });
+
+  it('is false for a Choice body', () => {
+    expect(isItemRestricted({ type: 'Choice', items: [] } as never)).toBe(
+      false
+    );
+  });
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:429]: That'd be an interesting one to dig deeper into maybe (note to self)

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 8]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+    });
+    const result =
+      getDownloadOptionsFromCanvasRenderingAndSupplementing(canvas);
+    expect(result.map(o => o.id)).toEqual([
+      'https://example.com/r.pdf',
+      'https://example.com/s.vtt',
+    ]);
+  });
+
+  it('flattens Choice bodies and drops string items', () => {
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts]: I'm not sure we see that in this example?

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 9]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+      }),
+    ];
+    expect(hasOriginalPdf(canvases)).toBe(true);
+  });
+
+  it('is false when there are no original PDFs', () => {
+    expect(hasOriginalPdf([createImageCanvas()])).toBe(false);
+    expect(hasOriginalPdf(undefined)).toBe(false);
+  });
+});
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:802]: Interesting to test it with a supplementing/painting PDF?

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 10]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+    const canvas = createImageCanvas({
+      painting: [],
+      supplementing: [
+        { id: 's', type: 'Text', format: 'application/pdf' },
+      ] as never,
+    });
+    expect(isPDFCanvas(canvas)).toBe(true);
+  });
+
+  it('is false for a PDF supplement when paintings are present (e.g. a video)', () => {
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:824]: This makes me wonder if `isPDFCanvas` is the right name for it, something to look into (how it's used), note for self.

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 11]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+
+  it('is false for an Image and undefined', () => {
+    expect(isAudioCanvas(createImageCanvas({ type: 'Image' as never }))).toBe(
+      false
+    );
+    expect(isAudioCanvas(undefined)).toBe(false);
+  });
+});
+
+describe('getItemsStatus', () => {
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:857]: Note to self, figure out what status and standards are...

File: content/webapp/utils/iiif/v3/index.test.ts [hunk 12]
```diff
@@ -350,3 +377,619 @@ describe('deduplicateDownloadOptions', () => {
...
+  it('labels a single specific type', () => {
+    expect(getFileTypeLabel(0, 2, false, [imageCanvas])).toBe('2 images');
+    expect(getFileTypeLabel(0, 1, false, [videoCanvas])).toBe('1 video file');
+    expect(getFileTypeLabel(0, 1, false, [audioCanvas])).toBe('1 audio file');
+  });
+
+  it('labels an empty canvas array as PDF files (vacuous every() quirk)', () => {
+    // With no canvases, `canvases.every(isPDFCanvas)` is vacuously true, so the
+    // all-PDFs branch is taken. Documenting current behaviour, not endorsing it.
+    expect(getFileTypeLabel(0, 0, false, [])).toBe('0 PDF files');
```
Comment (rcantin-w) [inline content/webapp/utils/iiif/v3/index.test.ts:924]: This is interesting, I don't _think_ this would be on purpose as it looks like image should be the default? (Should the default be undefined??)

It's not part of our mission now but I think worth raising for a potential fix.

**Review by rcantin-w — COMMENTED** [review]
This was such an educational read! Starting to think we should keep them forever as documentation 😄

File: content/webapp/test/fixtures/iiif/render.test.tsx (hunk 4 above)
Comment (rcantin-w) [reply content/webapp/test/fixtures/iiif/render.test.tsx:47]: Raph to dig into the why

File: content/webapp/utils/iiif/v3/canvas.test.ts (hunk 5 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/canvas.test.ts:89]: yes let's add one

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 8 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts]: We should review all `it`s and make sure the test actually covers them all.

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 9 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:802]: let's add

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 12 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:924]: I will raise with the team before we close this

File: content/webapp/test/fixtures/iiif/render.test.tsx (hunk 4 above)
Comment (rcantin-w) [reply content/webapp/test/fixtures/iiif/render.test.tsx:47]: Ok so the whole thing kind of confused me. This is a test within a test really (with the conditional), so that's why I wasn't convinced. Now I get it just tests if contexts get passed properly.

> This test file is a meta-test that verifies the `renderWithContext` test utility itself works correctly, rather than testing actual application features. It uses a Probe component (which only exists for testing) to read values from three React contexts and display them. The tests check that `renderWithContext` provides sensible defaults (like a single-canvas manifest), properly applies context overrides when you pass them in, and that the `createMockCanvas` factory works. 
> The confusing part is that it's using artificial test code to verify other test code, which can feel circular and low-value, real component tests would likely catch any issues with `renderWithContext` anyway, and the third test especially is just verifying that setting a property returns that same property, which is trivially true.

That being said, we are ALL up for over testing as part of this, it'll just be part of what goes I think, when we're done.

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 8 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts]: @davidpmccormick 's comment was for here, copying it in the right thread and deleting from the other:

 Three steps:
1. **Combine** `canvas.rendering` and `canvas.supplementing` into one list.
2. **Flatten Choice bodies.** In the IIIF spec, a `Choice` body represents *alternative renditions of the same content* (e.g. the same document as both a PDF and a Word doc). Rather than one downloadable item, a Choice wraps an `items` array of alternatives. `flatMap` unwraps it: if the item is a Choice, substitute its `items` array in place of the Choice itself; otherwise keep the item as a single-element array. So a Choice with 2 alternatives becomes 2 separate download candidates.
3. **Drop bare strings.** IIIF's `Choice.items` type technically permits array elements to be plain strings (a shorthand content-resource reference), not just full resource objects. The `filter` throws those away, since `convertToDownloadOption` needs an object with `.id`/`.label`/`.format`.

## What this specific test checks — and its actual gap

The test builds a `rendering` array containing one `Choice` whose `items` array has **one object** (a PDF). Running it through the function confirms the Choice gets unwrapped and its single alternative surfaces as a download option.

But looking at the fixture again: **there's no string in `items`** — only the one PDF object. So the test's name ("...and drops string items") claims to cover the string-filtering behaviour, but nothing in the test data actually exercises that branch. If someone broke the `.filter(item => typeof item !== 'string')` line entirely, this test would still pass, because there's no string for it to (fail to) drop.

That's worth fixing — I'd add a string entry alongside the object in `items` and assert it's excluded from the result. Want me to do that?

Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts]: @davidpmccormick I rewrote that test to only check the Choice flattening (by adding an item in that array), and added tests that check the empty string behaviour separately. https://github.com/wellcomecollection/wellcomecollection.org/pull/13249/changes/0fd8f5e0d248e241c7228241b6e692b56ab23857

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 9 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:802]: @davidpmccormick added https://github.com/wellcomecollection/wellcomecollection.org/pull/13249/changes/0fd8f5e0d248e241c7228241b6e692b56ab23857

File: content/webapp/utils/iiif/v3/canvas.test.ts (hunk 5 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/canvas.test.ts:89]: @davidpmccormick added https://github.com/wellcomecollection/wellcomecollection.org/pull/13249/changes/0fd8f5e0d248e241c7228241b6e692b56ab23857

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 7 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:429]: OK so for context above, I was wondering whether it was right to return false for a choice body in the `isItemRestricted` helper.

So as far as I can see, `isItemRestricted` is only used within `hasRestrictedItems`

```
// Returns true if any item in painting,
// original, or supplementing arrays is restricted.
export function hasRestrictedItem(canvas: TransformedCanvas): boolean {
  return (
    (canvas.painting?.some(item => isItemRestricted(item)) ?? false) ||
    (canvas.original?.some(original => isItemRestricted(original)) ?? false) ||
    (canvas.supplementing?.some(supp => isItemRestricted(supp)) ?? false)
  );
}
```

Makes me wonder if we 
1. need it as a separate helper (maybe, it looks tidier, but maybe the naming in confusing at a higher level, I feel like it'd be the helper to use to check a canvas)
2. if so, should we refactor it (it currently takes a param called `painting` when actually it takes `original` and `supplementing` as well - or is that just the name of that sort of object?
3. not wondering: we should type the expected param

The Choice body returning false is quite on purpose:
```
export function isItemRestricted(painting): boolean {
  if (isChoiceBody(painting)) return false;
  if (!painting.service) return false;

  const paintingsServices = painting.service.map(s => {
    if (s.type === 'AuthProbeService2') {
      return s.service.find(service => service.type === 'AuthAccessService2');
    }
    return undefined;
  });

  return paintingsServices?.some(s => {
    return (
      s?.id ===
      'https://iiif.wellcomecollection.org/auth/v2/access/restrictedlogin'
    );
  });
}
```

but I would like to understand why, maybe add a comment to that purpose. 

Claude seemed to think that because Choices themselves don't give information as to the restriction it was fine since Choices seem to be unwrapped at higher level and then potentially checked again, but we don't do that in the one place it's used, so I think this might be worth a chat!

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 10 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:824]: So definitely on purpose (I wrote that bit over a year ago aahaha)

```

  // 1. Born digital PDFs require a look at originals as their format is "Image"
  // 2. Because Videos could also have PDF supplements,
  //    we only want it to return true if it has no paintings.
  return hasOriginalPdf([canvas]) || (hasPDFSupplement && !hasPaintings);
  ```
  
  But yes it makes the name confusing. Some suggestions for the part where we rename things:
  > `isPDFCanvas` sounds like "does this canvas contain a PDF?" but it really means "should this canvas be treated as a PDF in the UI?" or "is the primary content a PDF?"
Better names might be:
> 
> 1. isPrimaryContentPDF
> 2. shouldTreatAsPDFCanvas
> 3. hasPDFAsPrimaryContent

Reckon 1 is good?

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 7 above)
Comment (davidpmccormick) [reply content/webapp/utils/iiif/v3/index.test.ts:429]: Will change the param name to `item` here, and type it properly in a future PR

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 12 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:924]: https://wellcome.slack.com/archives/CUA669WHH/p1784023441999599

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 10 above)
Comment (davidpmccormick) [reply content/webapp/utils/iiif/v3/index.test.ts:824]: Agree – let's rename in a separate PR

Comment (davidpmccormick) [reply content/webapp/utils/iiif/v3/index.test.ts:824]: #13270

File: content/webapp/utils/iiif/v3/index.test.ts (hunk 12 above)
Comment (rcantin-w) [reply content/webapp/utils/iiif/v3/index.test.ts:924]: https://github.com/wellcomecollection/wellcomecollection.org/issues/13271

**Review by rcantin-w — APPROVED** [review]
